### PR TITLE
Add ability to control number of fractional digits

### LINKED ISSUE #57: New option for precision
Would be good to have an option to allow one digit after the floating-point. 

Especially for scientific based apps.

### PR DESCRIPTION
**Pull request by nmoinvaz**
This should resolve #57.

### CODE REVIEW

**Comment by nmoinvaz** [comment]
I can remove fixed option if you don't want it.

File: readme.md [hunk 1]
```diff
@@ -72,6 +72,20 @@ Default: `false`
...
+
+Number of significant digits to display when calculating bytes or bits.
+
+##### fixed
+
+Type: `boolean`\
+Default: `false`
+
+Format the number using fixed-point notation, rounded up.
+
```
Comment (sindresorhus) [inline readme.md:88]: Add the example here, like in the TS file.

File: readme.md [hunk 2]
```diff
@@ -72,6 +72,20 @@ Default: `false`
 
 Format the number using the [Binary Prefix](https://en.wikipedia.org/wiki/Binary_prefix) instead of the [SI Prefix](https://en.wikipedia.org/wiki/SI_Prefix). This can be useful for presenting memory amounts. However, this should not be used for presenting file sizes.
 
+##### digits
+
+Type: `number`\
+Default: `3`
+
+Number of significant digits to display when calculating bytes or bits.
+
```
Comment (sindresorhus) [inline readme.md:81]: Add the example here, like in the TS file.

File: readme.md [hunk 3]
```diff
@@ -72,6 +72,20 @@ Default: `false`
...
+##### fixed
+
+Type: `boolean`\
+Default: `false`
+
+Format the number using fixed-point notation, rounded up.
```
Comment (sindresorhus) [inline readme.md:87]: From the description, it's not clear to a user what this actually does. I think it should be explained in simpler terms.

File: readme.md [hunk 4]
```diff
@@ -72,6 +72,20 @@ Default: `false`
 
 Format the number using the [Binary Prefix](https://en.wikipedia.org/wiki/Binary_prefix) instead of the [SI Prefix](https://en.wikipedia.org/wiki/SI_Prefix). This can be useful for presenting memory amounts. However, this should not be used for presenting file sizes.
 
+##### digits
+
+Type: `number`\
+Default: `3`
+
+Number of significant digits to display when calculating bytes or bits.
```
Comment (sindresorhus) [inline readme.md:80]: I don't think "calculating" is the correct word to use here. And does it only apply when showing bytes and not megabytes?

File: index.js [hunk 5]
```diff
@@ -79,8 +79,11 @@ module.exports = (number, options) => {
 
 	const exponent = Math.min(Math.floor(options.binary ? Math.log(number) / Math.log(1024) : Math.log10(number) / 3), UNITS.length - 1);
 	// eslint-disable-next-line unicorn/prefer-exponentiation-operator
-	number = Number((number / Math.pow(options.binary ? 1024 : 1000, exponent)).toPrecision(3));
-	const numberString = toLocaleString(number, options.locale);
+	number /= Math.pow(options.binary ? 1024 : 1000, exponent);
+	const digits = options.digits || 3;
```
Comment (sindresorhus) [inline index.js:83]: This makes it impossible to set it to `0`. I think handling the defaults should be done higher up where `Object.assign` is.

File: readme.md [hunk 6]
```diff
@@ -72,6 +72,20 @@ Default: `false`
 
 Format the number using the [Binary Prefix](https://en.wikipedia.org/wiki/Binary_prefix) instead of the [SI Prefix](https://en.wikipedia.org/wiki/SI_Prefix). This can be useful for presenting memory amounts. However, this should not be used for presenting file sizes.
 
+##### digits
```
Comment (sindresorhus) [inline readme.md:75]: The name here is too ambiguous. Digits could mean the non-fractional digits too.

File: index.d.ts [hunk 7]
```diff
@@ -48,6 +48,34 @@ declare namespace prettyBytes {
...
+		/**
+		Number of significant digits to display when calculating bytes or bits.
+
+		@default 3
+
+		```
+		import prettyBytes = require('pretty-bytes');
+
+		prettyBytes(1019, {digits: 3});
+		//=> '1.02 B'
```
Comment (sindresorhus) [inline index.d.ts:61]: The example should show with and without the option set so users can compare. Same with the other option.

File: index.d.ts [hunk 8]
```diff
@@ -48,6 +48,34 @@ declare namespace prettyBytes {
...
+
+		/**
+		Number of significant digits to display when calculating bytes or bits.
+
+		@default 3
+
+		```
+		import prettyBytes = require('pretty-bytes');
+
+		prettyBytes(1019, {digits: 3});
```
Comment (sindresorhus) [inline index.d.ts:60]: It makes no sense for the example to use the default value. That's just confusing.

File: index.d.ts [hunk 9]
```diff
@@ -48,6 +48,34 @@ declare namespace prettyBytes {
...
+		```
+		*/
+		readonly digits?: number;
+
+		/**
+		Format the number using fixed-point notation, rounded up.
+
+		@default false
+
+		```
```
Comment (sindresorhus) [inline index.d.ts:71]:
```suggestion
		@example
		```
```

File: test.js [hunk 10]
```diff
@@ -98,3 +98,21 @@ test('binary option', t => {
...
+});
+
+test('fixed option', t => {
+	t.is(prettyBytes(1001, {fixed: true}), '1.001 kB');
+	t.is(prettyBytes(1111, {fixed: true}), '1.111 kB');
+	t.is(prettyBytes(1900, {digits: 1, fixed: true}), '1.9 kB');
+	t.is(prettyBytes(1111, {digits: 2, fixed: true}), '1.11 kB');
+	t.is(prettyBytes(1019, {digits: 3, fixed: true}), '1.019 kB');
+	t.is(prettyBytes(1001, {digits: 4, fixed: true}), '1.001 kB');
+});
```
Comment (sindresorhus) [inline test.js:118]: Needs some more tests in combination with the other options.

**Review by sindresorhus — CHANGES_REQUESTED** [review]

**Comment by sindresorhus** [comment]
What's the use-case for the `fixed` option?

**Comment by nmoinvaz** [comment]
There was no use case for `fixed`, I just added it as an alternative to precision for completeness. I will remove it.

**Comment by nmoinvaz** [comment]
It appears the proper way to do this is not through `toFixed` to `toPrecision` but through locale options when calling `number.toLocaleString`. I can expose a `localeOptions` option which would allow users to set it through `localeOptions: {minimumFractionDigits: 2, maximumFractionDigits: 2}`, or I can keep with `digits: 2` with a note that using `digits` with no `locale` will automatically format the number using the system default locale.

File: readme.md (hunk 6 above)
Comment (nmoinvaz) [reply readme.md:75]: Do you prefer `fractionDigits`?

**Comment by nmoinvaz** [comment]
I have tried both ways, in two commits. I think the last commit is the best that has a `localeOptions` which is more configurable because both `minimumFractionDigits` and `maximumFractionDigits` can be set if desired.

File: index.d.ts [hunk 11]
```diff
@@ -19,6 +19,25 @@ declare namespace prettyBytes {
...
+		import prettyBytes = require('pretty-bytes');
+
+		prettyBytes(1900, {localeOptions: {maximumFractionDigits: 1}});
+		//=> '1.9 kB'
+
+		prettyBytes(1900, {localeOptions: {minimumFractionDigits: 3}});
+		//=> '1.900 kB'
+		```
+		*/
+		readonly localeOptions?: object;
```
Comment (sindresorhus) [inline index.d.ts:39]: `object` is not good enough. It needs to use a proper type.

**Comment by sindresorhus** [comment]
I agree with `minimumFractionDigits` and `maximumFractionDigits`, but they should be top-level options, not nested in `localOptions`.

**Comment by nmoinvaz** [comment]
The default values will be `undefined` because the when `minimumFractionDigits` and `maximumFractionDigits` is not set it will default to `toPrecision(3)` which has entirely different behavior.

**Comment by nmoinvaz** [comment]
Actually if you meant default value in the `Object.assign` for `options`, then I have included it.

**Comment by nmoinvaz** [comment]
Requested changes have been made again.

File: index.js [hunk 12]
```diff
@@ -70,7 +70,7 @@ module.exports = (number, options) => {
 		throw new TypeError(`Expected a finite number, got ${typeof number}: ${number}`);
 	}
 
-	options = Object.assign({bits: false, binary: false}, options);
+	options = Object.assign({bits: false, binary: false, minimumFractionDigits: undefined, maximumFractionDigits: undefined}, options);
```
Comment (sindresorhus) [inline index.js:73]: This is moot as unset properties are undefined.

File: readme.md [hunk 13]
```diff
@@ -85,6 +85,36 @@ Default: `false` *(No localization)*
...
+
+Type: `number`\
+Default: `undefined`
+
+The minimum number of fraction digits to display.
+
+```js
+const prettyBytes = require('pretty-bytes');
+
+// Show number with at least 3 fractional digits
```
Comment (sindresorhus) [inline readme.md:98]: index.d.ts and readme should be in sync.

File: index.d.ts [hunk 14]
```diff
@@ -51,6 +51,35 @@ declare namespace prettyBytes {
 		```
 		*/
 		readonly binary?: boolean;
+
+		/**
+		The minimum number of fraction digits to display.
```
Comment (sindresorhus) [inline index.d.ts:56]: It should say what the default behavior is (when not specified)

File: index.d.ts [hunk 15]
```diff
@@ -51,6 +51,35 @@ declare namespace prettyBytes {
...
+
+		@default undefined
+
+		```
+		import prettyBytes = require('pretty-bytes');
+
```
Comment (sindresorhus) [inline index.d.ts:62]: Also add an example of not using this option so the user can compare. Same with the other option.

File: index.js [hunk 16]
```diff
@@ -87,15 +87,30 @@ module.exports = (number, options) => {
...
+		localeOptions = {minimumFractionDigits: options.minimumFractionDigits};
+	}
+
+	if (options.maximumFractionDigits !== undefined) {
+		localeOptions = Object.assign({maximumFractionDigits: options.maximumFractionDigits}, localeOptions);
+	}
+
 	if (number < 1) {
-		const numberString = toLocaleString(number, options.locale);
+		const numberString = toLocaleString(number, options.locale, localeOptions);
```
Comment (sindresorhus) [inline index.js:101]:
```suggestion
		const numberString = toLocaleString(number, options.locale, {
			...{
				options.minimumFractionDigits,
				options.maximumFractionDigits
			}
		});
```

File: test.js [hunk 17]
```diff
@@ -113,3 +113,15 @@ test('bits and binary option', t => {
...
+test('fraction digits option', t => {
+	t.is(prettyBytes(1900, {maximumFractionDigits: 1}), '1.9 kB');
+	t.is(prettyBytes(1900, {minimumFractionDigits: 3}), '1.900 kB');
+	t.is(prettyBytes(1911, {maximumFractionDigits: 1}), '1.9 kB');
+	t.is(prettyBytes(1111, {maximumFractionDigits: 2}), '1.11 kB');
+	t.is(prettyBytes(1019, {maximumFractionDigits: 3}), '1.019 kB');
+	t.is(prettyBytes(1001, {maximumFractionDigits: 3}), '1.001 kB');
+	t.is(prettyBytes(4001, {maximumFractionDigits: 3, binary: true}), '3.907 kiB');
+	t.is(prettyBytes(18717, {maximumFractionDigits: 2, binary: true}), '18.28 kiB');
+	t.is(prettyBytes(18717, {maximumFractionDigits: 4, binary: true}), '18.2783 kiB');
```
Comment (sindresorhus) [inline test.js:126]: Needs some tests combining the two options.

**Review by sindresorhus — CHANGES_REQUESTED** [review]

File: index.js (hunk 16 above)
Comment (nmoinvaz) [reply index.js:101]: `localeOptions` should be `undefined` if both `minimumFractionDigits` and `maximumFractionDigits` is not set - so that it does not get localized by `toLocaleString`.

File: index.js (hunk 12 above)
Comment (nmoinvaz) [reply index.js]: Fixed.

File: index.js [hunk 18]
```diff
@@ -58,7 +58,7 @@ const toLocaleString = (number, locale, options) => {
 	let result = number;
```
[the remark this answers is not in the document]
Comment (nmoinvaz) [reply index.js:58]: Then I would check `Object.keys(options).length > 0` in the second if statement?

File: index.d.ts (hunk 14 above)
Comment (nmoinvaz) [reply index.d.ts:56]: Done.

File: index.d.ts (hunk 15 above)
Comment (nmoinvaz) [reply index.d.ts:62]: Done.

File: readme.md (hunk 13 above)
Comment (nmoinvaz) [reply readme.md:99]: Done.

File: test.js (hunk 17 above)
Comment (nmoinvaz) [reply test.js:128]: Done.

**Comment by nmoinvaz** [comment]
Some of the requested changes have been made.
